src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -46,22 +216,81 @@ class InvalidSchemaError(InvalidMessageError):
     pass
 
 
+class QuerySubscriptionStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def __init__(self, topic: str):
+        self.topic = topic
+
+    def create_with_partitions(
+        self,
+        commit: Commit,
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        def process_message(message: Message[KafkaPayload]) -> None:
+            with sentry_sdk.start_transaction(
+                op="handle_message",
+                name="query_subscription_consumer_process_message",
+                sampled=random() <= options.get("subscriptions-query.sample-rate"),
+            ), metrics.timer("snuba_query_subscriber.handle_message"):
+                value: Any = message.value
+                offset: int = value.offset
+                partition: int = value.partition.index
+                try:
+                    handle_message(
+                        str(message.payload.value),
+                        offset,
+                        partition,
+                        self.topic,
+                    )
+                except Exception:
+                    # This is a failsafe to make sure that no individual message will block this
+                    # consumer. If we see errors occurring here they need to be investigated to
+                    # make sure that we're not dropping legitimate messages.
+                    logger.exception(
+                        "Unexpected error while handling message in QuerySubscriptionStrategy. Skipping message.",
+                        extra={
+                            "offset": offset,
+                            "partition": partition,
+                            "value": message.value,
+                        },
+                    )
+
+        return RunTask(process_message, CommitOffsets(commit))
+
+
+def get_query_subscription_consumer(
+    topic: str,
+    group_id: str,
+    auto_offset_reset: str,
+    strict_offset_reset: bool,
+    force_topic: Union[str, None],
+    initial_offset_reset: str = "earliest",
+) -> StreamProcessor[KafkaPayload]:
+    topic = force_topic or topic
+    cluster_name = settings.KAFKA_TOPICS[topic]["cluster"]
+    cluster_options = kafka_config.get_kafka_consumer_cluster_options(cluster_name)
+    consumer = KafkaConsumer(
+        build_kafka_consumer_configuration(
+            cluster_options,
+            auto_offset_reset=auto_offset_reset,
+            group_id=group_id,
+            strict_offset_reset=strict_offset_reset,
+        )
+    )
+    return StreamProcessor(
```

Comment:
If you want to configure automatic metrics via Arroyo for this consumer you need to call `configure_metrics` with your metrics backend before this line. There is no docs page for this yet, but I wrote up some information about it here: https://github.com/getsentry/arroyo/pull/198